### PR TITLE
CDP improvment

### PR DESCRIPTION
**Pull request by mori-b**
CDP parsing and construction improvement : 
- create well formed packets (previous version had malformed packets visible in Wireshark)
- generic TLVs are supported automatically, special TLVs support is easier to add

### CODE REVIEW

**Comment by mori-b** [comment]
Curiously the coverage test fails only in HTTP for python 3.5, even though the only changes are relative to CDP. Moreover this specific failing test ".tox/py35/bin/py.test --cov=dpkt dpkt" works on my local machine.

**Comment by mori-b** [comment]
Second identical coverage test run passed ...

File: dpkt/cdp.py [hunk 1]
```diff
@@ -43,69 +44,111 @@ class CDP(dpkt.Packet):
...
         dpkt.Packet.unpack(self, buf)
         buf = self.data
         l = []
         while buf:
-            tlv = self.TLV(buf)
-            l.append(tlv)
+            #find the right TLV according to Type value
+            tlv_find_type = self.TLV(buf).type
+            #if this TLV is not in tlv_types, use the default TLV class
+            tlv = getattr(self, self.tlv_types.get(tlv_find_type, 'TLV'))(buf)
```
Comment (obormot) [inline dpkt/cdp.py]: Why getting class by name? It brings all this getattr() magic that seems unnecessary. You can just specify the actual class, e.g.
```
tlv_types = {CDP_ADDRESS: TLV_Addresses}
```

**Comment by mori-b** [comment]
@obormot 
Answering your remark regarding the use of getattr : I removed it in last commit, the downside is that then I have to define a class attribute at the end of the class, so that it can access the nested classes

File: dpkt/cdp.py (hunk 1 above)
Comment (mori-b) [inline dpkt/cdp.py]: Done as you suggested in last commit. Only downside is that the tlv_types class attribute must now be defined at the end of the class body, in order to access the nested classes.

**Comment by mori-b** [comment]
In spite of the fact I added tests to the CDP module, the automatic check shows a decrease of coverage. An explanation about its meaning could help, thanks.

**Comment by obormot** [comment]
This .01% is likely (unfortunately) due to this block not getting executed during the test:
```
if __name__ == '__main__':
    test_cdp()
    print('Tests Successful...')
```

You can click on "Details" right of the coverals to see the detailed report from coverals.

**Comment by brifordwylie** [comment]
@mori-b @obormot yeah.. lets not worry about the 0.01% coverage issue.. it's minutia... if the PR is good.. approve/merge it

**Comment by mori-b** [comment]
@obormot I see, the coverage test runs directly the "test_cdp" function, without needing to invoke the "if __name__=='__main__'" code. It seems to be the right way, it's the same behavior as Icmp.
So it looks like the coverage decreased because of useful code most of which is rightfully not in the scope of tox.

**Pull request merged**